### PR TITLE
First draft of a FIPS indicator design

### PR DESCRIPTION
**Pull request by hlandau**
In which I try to do FIPS things. 

I'm mostly braindumping this because my brain generated it automatically, and I thought it was good enough to serve as a starting point for discussion. I'm sure people will have much to say about it.

Part of https://github.com/openssl/project/issues/245

### CODE REVIEW

File: doc/designs/fips-indicator.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * indicator bit was reset were performed in compliance with the requirements
+ * of the given indicator bit. If a bit is cleared to 0, this means that at
+ * least one operation in violation of the requirements of that indicator bit
+ * has been performed since the indicators were reset.
+ *
+ * The indicator flags are thread-local, and therefore reflect operations
+ * performed on the current thread.
+ */
+uint64_t OPENSSL_get_fips_indicators(void);
+
```
Comment (nhorman) [inline doc/designs/fips-indicator.md:35]: Is a uint64_t going to be sufficiently large here? The granularity of our indicators may be such that we are going to have lots of them (several issues suggest that we may need an indicator for specific algorithms operating with specific modes/key sizes).  As such, as we move forward, we may well exceed 64 individual indicators

File: doc/designs/fips-indicator.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+/*
+ * Clears zero or more FIPS indicator bits, transitioning them to 0. This
+ * is to be used when an algorithm is executed which does not comply with
+ * the requirements of an indicator bit. This function is idempotent.
+ * The argument is a bitmask of the indicator bits which should be cleared.
+ */
+void OPENSSL_clear_fips_indicators(uint64_t indicators);
+
```
Comment (nhorman) [inline doc/designs/fips-indicator.md:43]: This should probably not be a public api call.  I don't think we want users defining fips compliance for certain indicators - i.e. indicators should be read-only for external library usage

**Comment by beldmit** [comment]
I think we should be ready to have more than one int64 values for indicators...

File: doc/designs/fips-indicator.md (hunk 2 above)
Comment (hlandau) [reply doc/designs/fips-indicator.md:43]: The idea is that providers can use it.

**Review by slontis — COMMENTED** [review]
Hmmm this is not what I was thinking at all :)
It really depends on what the purpose of this indicator really is.
I was thinking more of a gettable parameter per algorithm dispatch table..
i.e. You could query before you actually perform the operation.

The getter would maybe return 1 ? if there was not enough information to determine if the operation is approved. It would be up to the user to call this at the right time.
After all the required set up is done it would either return 1 or 0 depending on if the operation is approved.

e.g. If a particular algorithm allowed an unapproved parameter such as padding or digest, or bad keysize.

**Comment by slontis** [comment]
SO I guess mu question is, is this indicator actually useful for anything other than satisfying a FIPS requirement..
i.e. SHould the user be able to test if the operation they are about to do is approved BEFORE they do the operation?

**Comment by hlandau** [comment]
>If we were using global (thread local state), I imagine you would need to reset the bit at the start of any operation... So you would still need to do this per algorithm (So what advantage does this have over a getter?)

Not much different from the error stack. The idea is that most cryptographic applications are going to compose multiple primitives, so they can just do the test at the end and find out if anything was non-FIPS-compliant.

And yeah, the param-based solution is certainly a possibility. Idea here was something a bit less awkward to use.

**Comment by beldmit** [comment]
Red Hat has implemented parameter-based indication and didn't implement the check before the operation. I'm not sure that it's always doable without a complete setup and will not require API changes

**Comment by t8m** [comment]
I do not think supporting a "get indicator before the operation" is always possible and I do not even think it should be a goal.

**Comment by t8m** [comment]
Also, what is the "get indicator before operation is performed" good for? It isn't much better than getting it after anyway as the application actually has to use approved operations to get things done anyway. Perhaps the only use case would be if there were multiple options to choose from which were approved with some fips providers and unapproved with others and you would want to choose some based on which fips provider you have loaded. But that sounds to me like so niche use case that supporting it is not important enough.

**Comment by hlandau** [comment]
Dropping this from the project board as it's received some comments and isn't a current item.

**Comment by paulidale** [comment]
I don't think this meets the FIPS requirements.

**Pull request closed without merging**